src/transformers/generation_utils.py

```diff
@@ -573,13 +573,16 @@ def _get_logits_processor(
         num_beam_groups: int,
         diversity_penalty: float,
         remove_invalid_values: bool,
+        logits_processor: Optional[LogitsProcessorList],
     ) -> LogitsProcessorList:
         """
         This class returns a :obj:`~transformers.LogitsProcessorList` list object that contains all relevant
         :obj:`~transformers.LogitsProcessor` instances used to modify the scores of the language model head.
         """
-        processors = LogitsProcessorList()
-
+        if logits_processor is None:
```

Comment:
Here I also think we shouldn't even call the function if `logits_processor` is used. It makes our live much easier and the design a bit cleaner. 

What I don't like about the current design is that:
- if say both `forced_eos_token_id` is provided in `generate()` and as an input at the moment, then we have two `forced_eos_token_id` processors in the list which leads to weird behavior...

A first simple solution is to just not call this function IMO. We could always later adapt it for more advanced functionality if needed.

Reply:
One thing to keep in mind is that we essentially deactivate a whole bunch of options without much transparency about it so a user would need to know which options correspond to a `logits_processor`. E.g. I just had to look up if `temperature` would be affected if I added a `logits_processor`. Same with `stopping_criteria`.

The other way around it is more transparent I think if something goes wrong: Ah I passed `forced_eos_token_id` and a `logits_processor` doing the same thing, maybe that is not good.

For me the main use-case is to add a custom processer/criteria in addition to what I can already do with the kwargs. Otherwise, why go the extra step of using the arguably more complicated API if you can just pass it as an argument.

What do you think?

Reply:
That's an argument in favour of @Narsil's option 1.

Reply:
That's a good argument @lvwerra and I fully understand what you mean. 

I'm however really concerned about the complexity that option 1 adds for IMO very few use cases. Also from a backwards breaking point of view it's pretty much impossible to go from option 1 to option 2 in the future if this feature becomes more important where as it's much easier to go from option 2 to option 1 in the future.

> For me the main use-case is to add a custom processer/criteria in addition to what I can already do with the kwargs. Otherwise, why go the extra step of using the arguably more complicated API if you can just pass it as an argument.

Very good point and that's the big drawback of option 2. I believe that people having to use special logits processors are able to create it themselves. 

Ok, how about we do something in between option 1 and option 2 that doesn't create a crazy complex logic. 

If one passes a logits processor, we do the following:
1) just create the normal logits_processor that would have been created without passing one.
2) if any object of the passed logits processor is in the already created logits processor then we raise an error and tell the user which logits processor was created twice (and ideally which paramter has to be changed for this)
3) If there is no error make a union of the two lists and we're good to go.

Reply:
I like this compromise, what do you think @Narsil?

Reply:
Think we can go for the solution if you want @lvwerra :-)

Reply:
We can do that.